pkg/hooks/manager.go

```diff
@@ -0,0 +1,228 @@
+	"net"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	hooksInfo "kubevirt.io/kubevirt/pkg/hooks/info"
+	hooksV1alpha1 "kubevirt.io/kubevirt/pkg/hooks/v1alpha1"
+	"kubevirt.io/kubevirt/pkg/log"
+	domainSchema "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	virtwrapApi "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+)
+
+type callackClient struct {
+	SocketPath          string
+	Version             string
+	subsribedHookPoints []*hooksInfo.HookPoint
+}
+
+var manager *Manager
+var once sync.Once
+
+type Manager struct {
+	callbacksPerHookPoint map[string][]*callackClient
+}
+
+func GetManager() *Manager {
+	once.Do(func() {
+		manager = &Manager{callbacksPerHookPoint: make(map[string][]*callackClient)}
+	})
+	return manager
+}
+
+func (m *Manager) Collect(numberOfRequestedHookSidecars uint, timeout time.Duration) error {
+	callbacksPerHookPoint, err := collectSideCarSockets(numberOfRequestedHookSidecars, timeout)
+	if err != nil {
+		return err
+	}
+	log.Log.Info("Collected all requested hook sidecar sockets")
+
+	sortCallbacksPerHookPoint(callbacksPerHookPoint)
+	log.Log.Infof("Sorted all collected sidecar sockets per hook point based on their priority and name: %v", callbacksPerHookPoint)
+
+	m.callbacksPerHookPoint = callbacksPerHookPoint
+
+	return nil
+}
+
+// TODO: Handle sockets in parallel, when a socket appears, run a goroutine trying to read Info from it
+func collectSideCarSockets(numberOfRequestedHookSidecars uint, timeout time.Duration) (map[string][]*callackClient, error) {
+	callbacksPerHookPoint := make(map[string][]*callackClient)
+	processedSockets := make(map[string]bool)
+
+	timeoutCh := time.After(timeout)
+
+	for uint(len(processedSockets)) < numberOfRequestedHookSidecars {
+		sockets, err := ioutil.ReadDir(HookSocketsSharedDirectory)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, socket := range sockets {
+			select {
+			case <-timeoutCh:
+				return nil, fmt.Errorf("Failed to collect all expected sidecar hook sockets within given timeout")
+			default:
+				if _, processed := processedSockets[socket.Name()]; processed {
+					continue
+				}
+
+				callackClient, notReady, err := processSideCarSocket(HookSocketsSharedDirectory + "/" + socket.Name())
+				if notReady {
+					log.Log.Info("Sidecar server might not be ready yet, retrying in the next iteration")
+					continue
+				} else if err != nil {
+					log.Log.Reason(err).Infof("Failed to process sidecar socket: %s", socket.Name())
+					return nil, err
+				}
+
+				for _, subsribedHookPoint := range callackClient.subsribedHookPoints {
+					callbacksPerHookPoint[subsribedHookPoint.GetName()] = append(callbacksPerHookPoint[subsribedHookPoint.GetName()], callackClient)
+				}
+
+				processedSockets[socket.Name()] = true
+			}
+		}
+
+		time.Sleep(time.Second)
+	}
+
+	return callbacksPerHookPoint, nil
+}
+
+func processSideCarSocket(socketPath string) (*callackClient, bool, error) {
```

Comment:
Not sure if we want to pass the timeout here too.

Reply:
It contains only Info call, that should not take long. The timeout is there mainly to wait for the image to start. But I increased timeout of the OnDefineDomain call, in case sidecar does more than changing domxml.

I did not want to make the manager even more complicated with context timeout sharing.